File: tests/test_transform_effects.py
"""Transform effects class test harness."""

import unittest
from fractions import Fraction

import opentimelineio as otio
import opentimelineio.test_utils as otio_test_utils

class VideoScaleTests(unittest.TestCase, otio_test_utils.OTIOAssertions):
    def test_constructor(self):
        scale = otio.schema.VideoScale(
            name="ScaleIt",
            width=100,
            height=120,
            metadata={
                "foo": "bar"
            }
        )
        self.assertEqual(scale.width, 100)
        self.assertEqual(scale.height, 120)
        self.assertEqual(scale.name, "ScaleIt")
        self.assertEqual(scale.metadata, {"foo": "bar"})

    def test_eq(self):
        scale1 = otio.schema.VideoScale(
            name="ScaleIt",
            width=120,
            height=130,
            metadata={
                "foo": "bar"
            }
        )
        scale2 = otio.schema.VideoScale(
            name="ScaleIt",
            width=120,
            height=130,
            metadata={
                "foo": "bar"
            }
        )
        self.assertIsOTIOEquivalentTo(scale1, scale2)

    def test_serialize(self):
        scale = otio.schema.VideoScale(
            name="ScaleIt",
            width=130,
            height=140,
            metadata={
                "foo": "bar"
            }
        )
        encoded = otio.adapters.otio_json.write_to_string(scale)
        decoded = otio.adapters.otio_json.read_from_string(encoded)
        self.assertIsOTIOEquivalentTo(scale, decoded)

    def test_setters(self):
        scale = otio.schema.VideoScale(
            name="ScaleIt",
            width=140,
            height=150,
            metadata={
                "foo": "bar"
            }
        )
        self.assertEqual(scale.width, 140)
        scale.width = 100
        self.assertEqual(scale.width,100)
        self.assertEqual(scale.height, 150)
        scale.height = 100
        self.assertEqual(scale.height,100)

class VideoCropTests(unittest.TestCase, otio_test_utils.OTIOAssertions):
    def test_cons(self):
        crop = otio.schema.VideoCrop(
            name="CropIt",
            left=2,
            right=3,
            top=4,
            bottom=5,
            metadata={
                "baz": "qux"
            }
        )
        self.assertEqual(crop.left, 2)
        self.assertEqual(crop.right, 3)
        self.assertEqual(crop.top, 4)
        self.assertEqual(crop.bottom, 5)
        self.assertEqual(crop.name, "CropIt")
        self.assertEqual(crop.metadata, {"baz": "qux"})

    def test_eq(self):
        crop1 = otio.schema.VideoCrop(
            name="CropIt",
            left=2,
            right=3,
            top=4,
            bottom=5,
            metadata={
                "baz": "qux"
            }
        )
        crop2 = otio.schema.VideoCrop(
            name="CropIt",
            left=2,
            right=3,
            top=4,
            bottom=5,
            metadata={
                "baz": "qux"
            }
        )
        self.assertIsOTIOEquivalentTo(crop1, crop2)

    def test_serialize(self):
        crop = otio.schema.VideoCrop(
            name="CropIt",
            left=2,
            right=3,
            top=4,
            bottom=5,
            metadata={
                "baz": "qux"
            }
        )
        encoded = otio.adapters.otio_json.write_to_string(crop)
        decoded = otio.adapters.otio_json.read_from_string(encoded)
        self.assertIsOTIOEquivalentTo(crop, decoded)

    def test_setters(self):
        crop = otio.schema.VideoCrop(
            name="CropIt",
            left=2,
            right=3,
            top=4,
            bottom=5,
            metadata={
                "baz": "qux"
            }
        )
        self.assertEqual(crop.left, 2)
        crop.left = 1
        self.assertEqual(crop.left, 1)
        crop.right = 3
        self.assertEqual(crop.right, 3)
        crop.top = 4
        self.assertEqual(crop.top, 4)
        crop.bottom = 7
        self.assertEqual(crop.bottom, 7)

class VideoPositionTests(unittest.TestCase, otio_test_utils.OTIOAssertions):
    def test_constructor(self):
        position = otio.schema.VideoPosition(
            name="PositionIt",
            x=11,
            y=12,
            metadata={
                "alpha": "beta"
            }
        )
        self.assertEqual(position.x, 11)
        self.assertEqual(position.y, 12)
        self.assertEqual(position.name, "PositionIt")
        self.assertEqual(position.metadata, {"alpha": "beta"})

    def test_eq(self):
        pos1 = otio.schema.VideoPosition(
            name="PositionIt",
            x=11,
            y=12,
            metadata={
                "alpha": "beta"
            }
        )
        pos2 = otio.schema.VideoPosition(
            name="PositionIt",
            x=11,
            y=12,
            metadata={
                "alpha": "beta"
            }
        )
        self.assertIsOTIOEquivalentTo(pos1, pos2)

    def test_serialize(self):
        position = otio.schema.VideoPosition(
            name="PositionIt",
            x=11,
            y=12,
            metadata={
                "alpha": "beta"
            }
        )
        encoded = otio.adapters.otio_json.write_to_string(position)
        decoded = otio.adapters.otio_json.read_from_string(encoded)
        self.assertIsOTIOEquivalentTo(position, decoded)

    def test_setters(self):
        position = otio.schema.VideoPosition(
            name="PositionIt",
            x=11,
            y=12,
            metadata={
                "alpha": "beta"
            }
        )
        self.assertEqual(position.x, 11)
        position.x = 1
        self.assertEqual(position.x, 1)
        self.assertEqual(position.y, 12)
        position.y = 2
        self.assertEqual(position.y, 2)

class VideoRotateTests(unittest.TestCase, otio_test_utils.OTIOAssertions):
    def test_constructor(self):
        rotate = otio.schema.VideoRotate(
            name="RotateIt",
            angle=45.25,
            metadata={
                "rot": "val"
            }
        )
        self.assertEqual(rotate.angle,45.25)
        self.assertEqual(rotate.name, "RotateIt")
        self.assertEqual(rotate.metadata, {"rot": "val"})

    def test_eq(self):
        rot1 = otio.schema.VideoRotate(
            name="RotateIt",
            angle=45.25,
            metadata={
                "rot": "val"
            }
        )
        rot2 = otio.schema.VideoRotate(
            name="RotateIt",
            angle=45.25,
            metadata={
                "rot": "val"
            }
        )
        self.assertIsOTIOEquivalentTo(rot1, rot2)

    def test_serialize(self):
        rotate = otio.schema.VideoRotate(
            name="RotateIt",
            angle=45.25,
            metadata={
                "rot": "val"
            }
        )
        encoded = otio.adapters.otio_json.write_to_string(rotate)
        decoded = otio.adapters.otio_json.read_from_string(encoded)
        self.assertIsOTIOEquivalentTo(rotate, decoded)

    def test_setters(self):
        rotate = otio.schema.VideoRotate(
            name="RotateIt",
            angle=45.25,
            metadata={
                "rot": "val"
            }
        )
        self.assertEqual(rotate.angle, 45.25)
        rotate.angle = 90.0
        self.assertEqual(rotate.angle, 90.0)

class VideoRoundCornersTests(unittest.TestCase, otio_test_utils.OTIOAssertions):
    def test_constructor(self):
        rounded_corners = otio.schema.VideoRoundCorners(
            name="doRoundCorners",
            radius=10,
            metadata={
                "round": "corners"
            }
        )
        self.assertEqual(rounded_corners.radius, 10)
        self.assertEqual(rounded_corners.name, "doRoundCorners")
        self.assertEqual(rounded_corners.metadata, {"round": "corners"})

    def test_eq(self):
        rounded_corners1 = otio.schema.VideoRoundCorners(
            name="doRoundCorners",
            radius=10,
            metadata={
                "round": "corners"
            }
        )
        rounded_corners2 = otio.schema.VideoRoundCorners(
            name="doRoundCorners",
            radius=10,
            metadata={
                "round": "corners"
            }
        )
        self.assertIsOTIOEquivalentTo(rounded_corners1, rounded_corners2)

    def test_serialize(self):
        rounded_corners = otio.schema.VideoRoundCorners(
            name="doRoundCorners",
            radius=10,
            metadata={
                "round": "corners"
            }
        )
        encoded = otio.adapters.otio_json.write_to_string(rounded_corners)
        decoded = otio.adapters.otio_json.read_from_string(encoded)
        self.assertIsOTIOEquivalentTo(rounded_corners, decoded)

    def test_setters(self):
        rounded_corners = otio.schema.VideoRoundCorners(
            name="doRoundCorners",
            radius=10,
            metadata={
                "round": "corners"
            }
        )
        self.assertEqual(rounded_corners.radius, 10)
        rounded_corners.radius = 20
        self.assertEqual(rounded_corners.radius, 20)

class VideoFlipTests(unittest.TestCase, otio_test_utils.OTIOAssertions):
    def test_constructor(self):
        flip = otio.schema.VideoFlip(
            name="FlipIt",
            flip_horizontally=True,
            flip_vertically=False,
            metadata={
                "flip": "val"
            }
        )
        self.assertEqual(flip.flip_horizontally, True)
        self.assertEqual(flip.flip_vertically, False)
        self.assertEqual(flip.name, "FlipIt")
        self.assertEqual(flip.metadata, {"flip": "val"})

    def test_eq(self):
        flip1 = otio.schema.VideoFlip(
            name="FlipIt",
            flip_horizontally=True,
            flip_vertically=False,
            metadata={
                "flip": "val"
            }
        )
        flip2 = otio.schema.VideoFlip(
            name="FlipIt",
            flip_horizontally=True,
            flip_vertically=False,
            metadata={
                "flip": "val"
            }
        )
        self.assertIsOTIOEquivalentTo(flip1, flip2)

    def test_serialize(self):
        flip = otio.schema.VideoFlip(
            name="FlipIt",
            flip_horizontally=True,
            flip_vertically=False,
            metadata={
                "flip": "val"
            }
        )
        encoded = otio.adapters.otio_json.write_to_string(flip)
        decoded = otio.adapters.otio_json.read_from_string(encoded)
        self.assertIsOTIOEquivalentTo(flip, decoded)

    def test_setters(self):
        flip = otio.schema.VideoFlip(
            name="FlipIt",
            flip_horizontally=True,
            flip_vertically=False,
            metadata={
                "flip": "val"
            }
        )
        self.assertEqual(flip.flip_horizontally, True)
        flip.flip_horizontally = False
        self.assertEqual(flip.flip_horizontally, False)
        self.assertEqual(flip.flip_vertically, False)
        flip.flip_vertically = True
        self.assertEqual(flip.flip_vertically, True)

File: tests/test_transform_effects.cpp
#include "utils.h"

#include <opentimelineio/clip.h>
#include <opentimelineio/deserialization.h>
#include <opentimelineio/externalReference.h>
#include <opentimelineio/transformEffects.h>

namespace otime = opentime::OPENTIME_VERSION;
namespace otio  = opentimelineio::OPENTIMELINEIO_VERSION;

int
main(int argc, char** argv)
{
    Tests tests;
    tests.add_test("test_video_transform_read", [] {
        using namespace otio;

        otio::ErrorStatus              status;
        SerializableObject::Retainer<> so =
            SerializableObject::from_json_string(
                R"(
            {
                "OTIO_SCHEMA": "Clip.1",
                "media_reference": {
                    "OTIO_SCHEMA": "ExternalReference.1",
                    "target_url": "unit_test_url",
                    "available_range": {
                        "OTIO_SCHEMA": "TimeRange.1",
                        "duration": {
                            "OTIO_SCHEMA": "RationalTime.1",
                            "rate": 24,
                            "value": 8
                        },
                        "start_time": {
                            "OTIO_SCHEMA": "RationalTime.1",
                            "rate": 24,
                            "value": 10
                        }
                    }
                },
                "effects": [
                    {
                        "OTIO_SCHEMA": "VideoScale.1",
                        "name": "scale",
                        "width": 100,
                        "height": 120,
                        "effect_name": "VideoScale",
                        "enabled": true
                    },
                    {
                        "OTIO_SCHEMA": "VideoPosition.1",
                        "name": "position",
                        "x": 10,
                        "y": 20,
                        "effect_name": "VideoPosition",
                        "enabled": true
                    },
                    {
                        "OTIO_SCHEMA": "VideoRotate.1",
                        "name": "rotate",
                        "angle": 45.5,
                        "effect_name": "VideoRotate",
                        "enabled": true
                    },
                    {
                        "OTIO_SCHEMA": "VideoCrop.1",
                        "name": "crop",
                        "left": 5,
                        "right": 6,
                        "top": 7,
                        "bottom": 8,
                        "effect_name": "VideoCrop",
                        "enabled": true
                    },
                    {
                        "OTIO_SCHEMA": "VideoRoundedCorners.1",
                        "name": "roundedCorners",
                        "radius": 80,
                        "effect_name": "VideoRoundedCorners",
                        "enabled": true
                    },
                    {
                        "OTIO_SCHEMA": "VideoFlip.1",
                        "name": "flip",
                        "flip_horizontally": true,
                        "flip_vertically": false,
                        "effect_name": "VideoFlip",
                        "enabled": true
                    },
                    {
                        "OTIO_SCHEMA": "VideoMask.1",
                        "name": "mask",
                        "mask_type": "REMOVE",
                        "mask_url": "mask_url",
                        "effect_name": "VideoMaskRemove",
                        "enabled": true
                    },
                    {
                        "OTIO_SCHEMA": "VideoMask.1",
                        "name": "mask",
                        "mask_type": "REPLACE",
                        "mask_url": "mask_url",
                        "effect_name": "VideoMaskReplace",
                        "mask_replacement_url": "mask_replacement_url",
                        "enabled": true
                    },
                    {
                        "OTIO_SCHEMA": "VideoMask.1",
                        "name": "mask",
                        "mask_type": "BLUR",
                        "mask_url": "mask_url",
                        "effect_name": "VideoMaskBlur",
                        "blur_radius": 10.1,
                        "enabled": true
                    }
                ]
            })",
                &status);

        if (is_error(status))
            throw std::invalid_argument(status.details);

        const Clip* clip = dynamic_cast<const Clip*>(so.value);
        assertNotNull(clip);

        auto effects = clip->effects();
        assertEqual(effects.size(), 9);

        auto video_scale = dynamic_cast<const VideoScale*>(effects[0].value);
        assertNotNull(video_scale);
        assertEqual(video_scale->width(), 100);
        assertEqual(video_scale->height(), 120);

        auto video_position = dynamic_cast<const VideoPosition*>(effects[1].value);
        assertNotNull(video_position);
        assertEqual(video_position->x(), 10);
        assertEqual(video_position->y(), 20);

        auto video_rotate = dynamic_cast<const VideoRotate*>(effects[2].value);
        assertNotNull(video_rotate);
        assertEqual(video_rotate->angle(), 45.5);

        auto video_crop = dynamic_cast<const VideoCrop*>(effects[3].value);
        assertNotNull(video_crop);
        assertEqual(video_crop->left(), 5);
        assertEqual(video_crop->right(), 6);
        assertEqual(video_crop->top(), 7);
        assertEqual(video_crop->bottom(), 8);

        auto video_rounded_corners = dynamic_cast<const VideoRoundedCorners*>(effects[4].value);
        assertNotNull(video_rounded_corners);
        assertEqual(video_rounded_corners->radius(), 80);

        auto video_flip = dynamic_cast<const VideoFlip*>(effects[5].value);
        assertNotNull(video_flip);
        assertEqual(video_flip->flip_horizontally(), true);
        assertEqual(video_flip->flip_vertically(), false);

        auto video_mask_remove = dynamic_cast<const VideoMask*>(effects[6].value);
        assertNotNull(video_mask_remove);
        assertEqual(video_mask_remove->mask_type(), std::string(VideoMask::MaskType::remove));
        assertEqual(video_mask_remove->mask_url(), std::string("mask_url"));

        auto video_mask_replace = dynamic_cast<const VideoMask*>(effects[7].value);
        assertNotNull(video_mask_replace);
        assertEqual(video_mask_replace->mask_type(), std::string(VideoMask::MaskType::replace));
        assertEqual(video_mask_replace->mask_url(), std::string("mask_url"));
        assertEqual(video_mask_replace->mask_replacement_url().value(), std::string("mask_replacement_url"));

        auto video_mask_blur = dynamic_cast<const VideoMask*>(effects[8].value);
        assertNotNull(video_mask_blur);
        assertEqual(video_mask_blur->mask_type(), std::string(VideoMask::MaskType::blur));
        assertEqual(video_mask_blur->mask_url(), std::string("mask_url"));
        assertEqual(video_mask_blur->blur_radius().value(), 10.1);
    });

    tests.add_test("test_video_transform_write", [] {
        using namespace otio;

        SerializableObject::Retainer<otio::Clip> clip(new otio::Clip(
            "unit_clip",
            new otio::ExternalReference("unit_test_url"),
            std::nullopt,
            otio::AnyDictionary(),
            { new otio::VideoScale("scale", 100, 120),
              new otio::VideoPosition("position", 10, 20),
              new otio::VideoRotate("rotate", 40.5),
              new otio::VideoCrop("crop", 1, 2, 3, 4),
              new otio::VideoRoundedCorners("roundedCorners",80),
              new otio::VideoFlip("flip", true, false),
              new otio::VideoMask("mask", otio::VideoMask::MaskType::remove, "mask_url")
            }));

        auto json = clip.value->to_json_string();

        std::string expected_json = R"({
    "OTIO_SCHEMA": "Clip.2",
    "metadata": {},
    "name": "unit_clip",
    "source_range": null,
    "effects": [
        {
            "OTIO_SCHEMA": "VideoScale.1",
            "metadata": {},
            "name": "scale",
            "effect_name": "VideoScale",
            "enabled": true,
            "width": 100,
            "height": 120
        },
        {
            "OTIO_SCHEMA": "VideoPosition.1",
            "metadata": {},
            "name": "position",
            "effect_name": "VideoPosition",
            "enabled": true,
            "x": 10,
            "y": 20
        },
        {
            "OTIO_SCHEMA": "VideoRotate.1",
            "metadata": {},
            "name": "rotate",
            "effect_name": "VideoRotate",
            "enabled": true,
            "angle": 40.5
        },
        {
            "OTIO_SCHEMA": "VideoCrop.1",
            "metadata": {},
            "name": "crop",
            "effect_name": "VideoCrop",
            "enabled": true,
            "left": 1,
            "right": 2,
            "top": 3,
            "bottom": 4
        },
        {
            "OTIO_SCHEMA": "VideoRoundedCorners.1",
            "metadata": {},
            "name": "roundedCorners",
            "effect_name": "VideoRoundedCorners",
            "enabled": true,
            "radius": 80
        },
        {
            "OTIO_SCHEMA": "VideoFlip.1",
            "metadata": {},
            "name": "flip",
            "effect_name": "VideoFlip",
            "enabled": true,
            "flip_horizontally": true,
            "flip_vertically": false
        },
        {
            "OTIO_SCHEMA": "VideoMask.1",
            "metadata": {},
            "name": "mask",
            "effect_name": "VideoMask",
            "enabled": true,
            "mask_type": "REMOVE",
            "mask_url": "mask_url"
        }
    ],
    "markers": [],
    "enabled": true,
    "media_references": {
        "DEFAULT_MEDIA": {
            "OTIO_SCHEMA": "ExternalReference.1",
            "metadata": {},
            "name": "",
            "available_range": null,
            "available_image_bounds": null,
            "target_url": "unit_test_url"
        }
    },
    "active_media_reference_key": "DEFAULT_MEDIA"
})";


        assertEqual(json, expected_json);

    });

    tests.run(argc, argv);
    return 0;
}
